src/library_webgl.js

```diff
@@ -3594,6 +3594,59 @@ var LibraryGL = {
     GLctx.sampleCoverage(value, !!invert);
   },
 
+  glMultiDrawArraysWEBGL: function(mode, firsts, firstsOffset, counts, countsOffset, drawcount) {
+    var firstsList = new Int32Array(HEAPU8.subarray(firsts, firsts + drawcount * 4));
```

Comment:
is the `new Int32Array()` call needed? That creates a copy, and I'm not sure why.

Reply:
Is there a way to access the heap as an Int32Array?

Reply:
Yes, `HEAP32` is a view with that type. So I think `HEAP32.subarray(first >> 2, first + drawcount * 4 >> 2)` would work?

Reply:
Looking at https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html?highlight=subarray#access-memory-from-javascript, my understanding is that if there's `-s ALLOW_MEMORY_GROWTH=1`, and the `firsts` pointer could become invalid?

Reply:
These pointers are just integer offsets into the WASM heap - they don't become invalid. 

I'm not 100% sure how memory growth works but the example in https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html?highlight=subarray#access-memory-from-javascript looks broken to me - the copy's being made out of the already-potentially-invalidated HEAP8.

I think you should ignore that document and just use the HEAP32 view as @kripken suggests above.

Reply:
Is the view retained by the called API? AFAIK old WebGL APIs don't do that, so growth is not a concern. But if this one does save the view, then it could be invalidated, and a copy is necessary.

Reply:
@kenrussell about the example, I think it's valid: malloc may invalidate `HEAPU8`, but we automatically recreate those global views when that happens. So the only risk is user views that we do not handle, that users must update themselves.

Reply:
Seems a copy would happen in the WebGL API impl. So creating a copy view here seems unnecessary to me
https://source.chromium.org/chromium/chromium/src/+/master:gpu/command_buffer/service/multi_draw_manager.cc;l=85?originalUrl=https:%2F%2Fcs.chromium.org%2F

Reply:
@kripken OK, thanks - the example's still a little confusing because it's the *next* call to func() which will invalidate the view returned from an earlier call.

@shrekshao agree, the copy is unnecessary.